pkg/kubelet/cm/internal_container_lifecycle_windows.go

```diff
@@ -20,30 +20,160 @@ limitations under the License.
+		// Case 3: CPU manager selects fewer CPUs, but some are outside of the CPUs available in the NUMA nodes selected by the memory manager
 
-			containerConfig.Windows.Resources.AffinityCpus = cpuGroupAffinities
+		if len(cpuManagerAffinityCPUSet) > len(numaNodeAffinityCPUSet) {
+			// Case 1, use CPU manager selected CPUs
+			finalCPUSet = cpuManagerAffinityCPUSet
+		} else if subset(cpuManagerAffinityCPUSet, numaNodeAffinityCPUSet) {
+			// case 2, use CPU manager selected CPUs
+			finalCPUSet = cpuManagerAffinityCPUSet
+		} else {
+			// Case 3, merge CPU manager and memory manager selected CPUs
+			finalCPUSet = mergeSets(cpuManagerAffinityCPUSet, numaNodeAffinityCPUSet)
+		}
+	} else if i.cpuManager != nil && !allocatedCPUs.IsEmpty() {
+		// Only CPU manager is enabled, use CPU manager selected CPUs
+		finalCPUSet = computeCPUSet(convertToGroupAffinities(allocatedCPUs))
+	} else if i.memoryManager != nil && !allocatedCPUs.IsEmpty() {
+		// Only memory manager is enabled, use CPUs associated with selected NUMA nodes
+		var allNumaNodeCPUs []winstats.GroupAffinity
+		for _, numaNode := range sets.List(numaNodes) {
+			affinity, err := winstats.GetCPUsforNUMANode(uint16(numaNode))
+			if err != nil {
+				return fmt.Errorf("failed to get CPUs for NUMA node %d: %v", numaNode, err)
+			}
+			allNumaNodeCPUs = append(allNumaNodeCPUs, *affinity)
 		}
+		finalCPUSet = computeCPUSet(allNumaNodeCPUs)
 	}
+
+	// Set CPU group affinities in the container config
+	if finalCPUSet != nil {
+		var cpusToGroupAffinities []*runtimeapi.WindowsCpuGroupAffinity
+		for group, mask := range groupMasks(finalCPUSet) {
+			cpusToGroupAffinities = append(cpusToGroupAffinities, &runtimeapi.WindowsCpuGroupAffinity{
+				CpuGroup: uint32(group),
+				CpuMask:  uint64(mask),
+			})
+		}
+		containerConfig.Windows.Resources.AffinityCpus = cpusToGroupAffinities
+	}
+
+	// return nil if no CPUs were selected
 	return nil
 }
+
+// computeCPUSet returns a map of CPU IDs to an empty struct based on the provided group affinities
+func computeCPUSet(affinities []winstats.GroupAffinity) map[int]struct{} {
+	cpuSet := make(map[int]struct{})
+	for _, affinity := range affinities {
+		for processorID := range affinity.Processors() {
+			cpuSet[processorID] = struct{}{}
+		}
+	}
+	return cpuSet
+}
+
+// subset returns true if set1 is a subset of set2
+func subset(set1, set2 map[int]struct{}) bool {
+	for k := range set1 {
+		if _, ok := set2[k]; !ok {
+			return false
+		}
+	}
+	return true
+}
+
+// mergeSets combines two sets of CPU IDs
+func mergeSets(set1, set2 map[int]struct{}) map[int]struct{} {
+	mergedSet := make(map[int]struct{})
+	for k := range set1 {
+		mergedSet[k] = struct{}{}
+	}
+	for k := range set2 {
+		mergedSet[k] = struct{}{}
+	}
+	return mergedSet
+}
+
+// convertToGroupAffinities converts a cpuset.CPUSet to a slice of winstats.GROUP_AFFINITY
+func convertToGroupAffinities(cpuSet cpuset.CPUSet) []winstats.GroupAffinity {
```

Comment:
This looks pretty close to https://github.com/kubernetes/kubernetes/blob/3140807126c86163584974e8a0835c2f5907af97/pkg/kubelet/winstats/cpu_topology.go#L91  

Could we provide another helper function and/or adjust the logic to convert to affinities?  My concern would be spreading the logic for calculating back and forth between kubelet's representation  (integers for the CPU's) to Windows representation of groupAffinities.

Reply:
I refactored some of the logic and got rid of the need for this method